engine/src/main/java/com/datatorrent/stram/util/PubSubWebSocketClient.java

```diff
@@ -239,17 +239,18 @@ public void onOpen(WebSocket ws)
     }
   }
 
+  public boolean isConnectionSetup()
```

Comment:
I don't see why is `isConnectionSetup()` necessary. Shared in SharedPubSubWebSocket client means it is shareable and must support concurrency. It should be the responsibility of SharedPubSubWebSocket to ensure that its state is properly managed and if `setup()` can't be called concurrently, enforce it.

Reply:
I see.. let me see what can be done. This thing keeps getting bigger.

Reply:
One thing though, shared doesn't necessarily mean it should be able to handle simultaneous initialization calls. A shareable resource can still ask for its initialization to be only called once and from one place. The rest of it is thread safe.

Reply:
Why should it be able to handle concurrent publish requests and not initialization requests?  In any case, why not to move that logic to `setup` and hide the complexity for all clients?

Reply:
Setup is a one-time activity but publish is not.

Reply:
I think it should not be called setup, because openConnection can be called multiple times, for example when connection wasn't successful before. We can just call it openConnection.